File: benchmarks/pandas/bench_sample_fn.py
"""
Benchmark: pandas sample on Series and DataFrame.
Outputs JSON: {"function": "sample_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

ROWS = 100_000
WARMUP = 3
ITERATIONS = 20

data = [i * 1.5 for i in range(ROWS)]
s = pd.Series(data)
df = pd.DataFrame({"a": data, "b": [x * 2 for x in data], "c": [x + 100 for x in data]})

for _ in range(WARMUP):
    s.sample(n=1000)
    s.sample(frac=0.01)
    df.sample(n=500)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.sample(n=1000)
    s.sample(frac=0.01)
    df.sample(n=500)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
mean_ms = total_ms / ITERATIONS
print(json.dumps({"function": "sample_fn", "mean_ms": round(mean_ms, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))

File: benchmarks/tsb/bench_sample_fn.ts
/**
 * Benchmark: sampleSeries / sampleDataFrame — standalone functional sample.
 * Outputs JSON: {"function": "sample_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, DataFrame, sampleSeries, sampleDataFrame } from "../../src/index.ts";

const ROWS = 100_000;
const WARMUP = 3;
const ITERATIONS = 20;

const data = Array.from({ length: ROWS }, (_, i) => i * 1.5);
const s = new Series({ data });
const df = new DataFrame(
  new Map([
    ["a", new Series({ data })],
    ["b", new Series({ data: data.map((x) => x * 2) })],
    ["c", new Series({ data: data.map((x) => x + 100) })],
  ]),
);

for (let i = 0; i < WARMUP; i++) {
  sampleSeries(s, { n: 1000 });
  sampleSeries(s, { frac: 0.01 });
  sampleDataFrame(df, { n: 500 });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  sampleSeries(s, { n: 1000 });
  sampleSeries(s, { frac: 0.01 });
  sampleDataFrame(df, { n: 500 });
  times.push(performance.now() - t0);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "sample_fn",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
